File: doctor-analyzer/backend/domain/analysis.py
"""Analysis result domain models."""

from dataclasses import dataclass, field
from typing import List, Optional, Dict, Any
from datetime import datetime
from enum import Enum


class AnalysisStatus(str, Enum):
    """Analysis job status."""

    PENDING = "pending"
    UPLOADING = "uploading"
    PROCESSING_VIDEO = "processing_video"
    PROCESSING_SELF_INJURY = "processing_self_injury"
    PROCESSING_AUDIO = "processing_audio"
    PROCESSING_BEDROCK = "processing_bedrock"
    AGGREGATING = "aggregating"
    COMPLETED = "completed"
    FAILED = "failed"


@dataclass(frozen=True)
class SentimentResult:
    """Sentiment analysis result from Comprehend."""

    sentiment: str  # POSITIVE, NEGATIVE, NEUTRAL, MIXED
    positive_score: float
    negative_score: float
    neutral_score: float
    mixed_score: float
    source_text: str = ""

    def to_dict(self) -> dict:
        return {
            "sentiment": self.sentiment,
            "positive_score": self.positive_score,
            "negative_score": self.negative_score,
            "neutral_score": self.neutral_score,
            "mixed_score": self.mixed_score,
        }


@dataclass(frozen=True)
class TranscriptionSegment:
    """Transcription segment with timing."""

    text: str
    start_time: float  # seconds
    end_time: float
    confidence: float
    speaker_label: Optional[str] = None

    def to_dict(self) -> dict:
        return {
            "text": self.text,
            "start_time": self.start_time,
            "end_time": self.end_time,
            "confidence": self.confidence,
            "speaker_label": self.speaker_label,
        }


@dataclass
class AudioAnalysis:
    """Complete audio analysis result."""

    transcription: List[TranscriptionSegment] = field(default_factory=list)
    overall_sentiment: Optional[SentimentResult] = None
    segment_sentiments: List[SentimentResult] = field(default_factory=list)

    @property
    def full_transcript(self) -> str:
        return " ".join(seg.text for seg in self.transcription)

    def to_dict(self) -> dict:
        return {
            "full_transcript": self.full_transcript,
            "segment_count": len(self.transcription),
            "overall_sentiment": self.overall_sentiment.to_dict() if self.overall_sentiment else None,
        }


@dataclass
class ClinicalIndicator:
    """Clinical indicator derived from analysis."""

    indicator_type: str  # discomfort, depression, anxiety, fear
    confidence: float
    evidence: List[str] = field(default_factory=list)
    timestamp_ranges: List[Dict] = field(default_factory=list)

    def to_dict(self) -> dict:
        return {
            "indicator_type": self.indicator_type,
            "confidence": self.confidence,
            "evidence": self.evidence,
            "timestamp_ranges": self.timestamp_ranges,
        }


@dataclass
class SelfInjuryCheckResult:
    """Result of optional self-injury check (Rekognition content moderation + optional Bedrock enhancement)."""

    enabled: bool
    rekognition_labels: List[Dict[str, Any]]  # name, confidence, timestamp_ms, parent_name
    has_signals: bool  # derived from labels (e.g. Self-Harm, Violence above threshold)
    summary: str
    confidence: float  # max confidence of relevant labels or 0
    error_message: Optional[str] = None
    severity: Optional[str] = None  # low/moderate/high/critical (from Bedrock)
    clinical_rationale: Optional[str] = None  # from Bedrock
    transcript_analysis: Optional[Dict[str, Any]] = None  # from Bedrock verbal analysis

    def to_dict(self) -> dict:
        base = {
            "enabled": self.enabled,
            "rekognition_labels": self.rekognition_labels,
            "bedrock_has_signals": self.has_signals,
            "bedrock_summary": self.summary,
            "bedrock_confidence": self.confidence,
            "error_message": self.error_message,
        }
        if self.severity is not None:
            base["bedrock_severity"] = self.severity
        if self.clinical_rationale is not None:
            base["bedrock_clinical_rationale"] = self.clinical_rationale
        if self.transcript_analysis is not None:
            base["bedrock_transcript_analysis"] = self.transcript_analysis
        return base

File: doctor-analyzer/frontend/src/types/analysis.ts
/**
 * Analysis type definitions matching backend domain models.
 */

export type AnalysisStatus =
  | 'pending'
  | 'uploading'
  | 'processing_video'
  | 'processing_injury_check'
  | 'processing_audio'
  | 'processing_bedrock'
  | 'aggregating'
  | 'completed'
  | 'failed';

export type EmotionType =
  | 'discomfort'
  | 'depression'
  | 'anxiety'
  | 'fear'
  | 'calm'
  | 'happy'
  | 'confused'
  | 'angry'
  | 'surprised'
  | 'disgusted'
  | 'sad'
  | 'neutral';

export interface EmotionScore {
  emotion: EmotionType;
  confidence: number;
}

export interface BoundingBox {
  left: number;
  top: number;
  width: number;
  height: number;
}

export interface FaceDetection {
  timestamp_ms: number;
  bounding_box: BoundingBox;
  emotions: EmotionScore[];
  primary_emotion: EmotionScore | null;
  age_range: { low: number; high: number } | null;
  gender: string | null;
}

export interface VideoEmotionTimeline {
  video_id: string;
  duration_ms: number;
  detection_count: number;
  emotion_summary: Record<string, number>;
}

export interface SentimentResult {
  sentiment: 'POSITIVE' | 'NEGATIVE' | 'NEUTRAL' | 'MIXED';
  positive_score: number;
  negative_score: number;
  neutral_score: number;
  mixed_score: number;
}

export interface TranscriptionSegment {
  text: string;
  start_time: number;
  end_time: number;
  confidence: number;
  speaker_label: string | null;
}

export interface AudioAnalysis {
  full_transcript: string;
  segment_count: number;
  overall_sentiment: SentimentResult | null;
}

export interface ClinicalIndicator {
  indicator_type: string;
  confidence: number;
  evidence: string[];
  timestamp_ranges: Array<{ start: number; end: number }>;
}

export interface ModerationLabelItem {
  name: string;
  confidence: number;
  timestamp_ms?: number;
  parent_name?: string;
}

export type InjurySeverity = 'low' | 'moderate' | 'high' | 'critical';

export interface TranscriptInjuryAnalysis {
  has_verbal_signals: boolean;
  severity: string;
  findings: string[];
  evidence_quotes: string[];
  confidence: number;
  risk_factors_identified?: string[];
}

export interface InjuryCheckResult {
  enabled: boolean;
  rekognition_labels: ModerationLabelItem[];
  bedrock_has_signals: boolean;
  bedrock_summary: string;
  bedrock_confidence: number;
  bedrock_severity?: InjurySeverity;
  bedrock_clinical_rationale?: string;
  bedrock_transcript_analysis?: TranscriptInjuryAnalysis;
  error_message?: string;
}

export interface BedrockAggregation {
  clinical_summary: string;
  risk_level: InjurySeverity;
  cross_referenced_evidence: Array<{
    source: string;
    finding: string;
    supporting_sources: string[];
  }>;
  concordant_signals: string[];
  discordant_signals: string[];
  recommendations: string[];
}

export interface Patient {
  id: string;
  codename: string;
  created_at: string;
}

export interface AnalysisSession {
  session_id: string;
  patient_id: string;
  created_at: string;
  updated_at: string | null;
  status: AnalysisStatus;
  video_s3_key: string | null;
  emotion_summary: Record<string, number>;
  clinical_indicators: ClinicalIndicator[];
  error_message: string | null;
  injury_check?: InjuryCheckResult | null;
  bedrock_aggregation?: BedrockAggregation | null;
}

export interface AnalysisSessionFull extends AnalysisSession {
  video_emotions: VideoEmotionTimeline | null;
  audio_analysis: AudioAnalysis | null;
  results_s3_key: string | null;
  injury_check: InjuryCheckResult | null;
  bedrock_aggregation: BedrockAggregation | null;
}

// WebSocket message types
export interface EmotionUpdateMessage {
  type: 'emotion_update';
  timestamp_ms: number;
  emotions: EmotionScore[];
  bounding_box: BoundingBox;
}

export interface StatusUpdateMessage {
  type: 'status_update';
  status: AnalysisStatus;
  progress?: number;
  message?: string;
}

export interface TranscriptionUpdateMessage {
  type: 'transcription_update';
  text: string;
  start_time: number;
  end_time: number;
}

export interface CompleteMessage {
  type: 'complete';
  results: AnalysisSessionFull;
}

export interface ErrorMessage {
  type: 'error';
  message: string;
}

export type WebSocketMessage =
  | EmotionUpdateMessage
  | StatusUpdateMessage
  | TranscriptionUpdateMessage
  | CompleteMessage
  | ErrorMessage;

// Upload response types
export interface UploadVideoResponse {
  session_id: string;
  video_s3_key: string;
  status: AnalysisStatus;
}

